app/main.py

```diff
@@ -1,214 +1,129 @@
-        "name": "Red Knight",
-        "power": 40,
-        "hp": 70,
-        "armour": [
-            {
-                "part": "breastplate",
-                "protection": 25,
-            }
-        ],
-        "weapon": {
-            "name": "Sword",
-            "power": 45
-        },
-        "potion": {
-            "name": "Blessing",
-            "effect": {
-                "hp": +10,
-                "power": +5,
-            }
+class Knight:
+    def __init__(self, name, power, hp, armour=None, weapon=None, potion=None):
+        self.name = name
+        self.base_power = power
+        self.base_hp = hp
+        self.armour = armour if armour else []
+        self.weapon = weapon
+        self.potion = potion
+        self.final_stats = self._calculate_stats()
+
+    def _calculate_stats(self):
+        final_power = self.base_power
+        final_hp = self.base_hp
+        final_protection = 0
+
+        # Додаємо захист від броні
+        for piece in self.armour:
+            final_protection += piece.get("protection", 0)
+
+        # Додаємо силу від зброї
+        if self.weapon:
+            final_power += self.weapon.get("power", 0)
+
+        # Додаємо ефекти зілля
+        if self.potion:
+            effect = self.potion.get("effect", {})
+            final_power += effect.get("power", 0)
+            final_hp += effect.get("hp", 0)
+            final_protection += effect.get("protection", 0)
+
+        return {
+            "power": final_power,
+            "hp": final_hp,
+            "protection": final_protection,
         }
-    }
-}
-
-
-def battle(knightsConfig):
-    # BATTLE PREPARATIONS:
-
-    # lancelot
-    lancelot = knightsConfig["lancelot"]
-
-    # apply armour
-    lancelot["protection"] = 0
-    for a in lancelot["armour"]:
-        lancelot["protection"] += a["protection"]
-
-    # apply weapon
-    lancelot["power"] += lancelot["weapon"]["power"]
-
-    # apply potion if exist
-    if lancelot["potion"] is not None:
-        if "power" in lancelot["potion"]["effect"]:
-            lancelot["power"] += lancelot["potion"]["effect"]["power"]
-
-        if "protection" in lancelot["potion"]["effect"]:
-            lancelot["protection"] += lancelot["potion"]["effect"]["protection"]
-
-        if "hp" in lancelot["potion"]["effect"]:
-            lancelot["hp"] += lancelot["potion"]["effect"]["hp"]
-
-    # arthur
-    arthur = knightsConfig["arthur"]
-
-    # apply armour
-    arthur["protection"] = 0
-    for a in arthur["armour"]:
-        arthur["protection"] += a["protection"]
-
-    # apply weapon
-    arthur["power"] += arthur["weapon"]["power"]
 
-    # apply potion if exist
-    if arthur["potion"] is not None:
-        if "power" in arthur["potion"]["effect"]:
-            arthur["power"] += arthur["potion"]["effect"]["power"]
+    def attack(self, other):
+        damage = max(0, self.final_stats["power"] - other.final_stats["protection"])
+        other.final_stats["hp"] -= damage
```

Comment:
Directly modifying the `final_stats` dictionary can lead to unexpected behavior if the same knight instance is used in multiple battles. Consider creating a method to handle HP reduction to ensure immutability and better encapsulation.